packages/cosmic-swingset/lib/block-manager.js

```diff
@@ -0,0 +1,177 @@
+const BEGIN_BLOCK = 'BEGIN_BLOCK';
+const DELIVER_INBOUND = 'DELIVER_INBOUND';
+const END_BLOCK = 'END_BLOCK';
+const COMMIT_BLOCK = 'COMMIT_BLOCK';
+
+export default function makeBlockManager({
+  deliverInbound,
+  beginBlock,
+  saveChainState,
+  saveOutsideState,
+  savedActions,
+  savedHeight,
+}) {
+  let computedHeight = savedHeight;
+  let runTime = 0;
+  async function kernelPerformAction(action) {
+    const start = Date.now();
+    const finish = _ => (runTime += Date.now() - start);
+
+    let p;
+    switch (action.type) {
+      case BEGIN_BLOCK:
+        p = beginBlock(action.blockHeight, action.blockTime);
+        break;
+
+      case DELIVER_INBOUND:
+        p = deliverInbound(
+          action.peer,
+          action.messages,
+          action.ack,
+          action.blockHeight,
+          action.blockTime,
+        );
+        break;
+
+      case END_BLOCK:
+        return true;
+
+      default:
+        throw new Error(`${action.type} not recognized`);
+    }
+    p.then(finish, finish);
+    return p;
+  }
+
+  let currentActions;
+  let currentIndex;
+  let replaying;
+  let decohered;
+  let saveTime = 0;
+
+  async function blockManager(action) {
```

Comment:
Overall structure looks good, and I know what the `blockManager` is supposed to be doing, but I have to admit I find this control flow hard to follow. We pass multiple times through the same function in different modes, with some code shared between modes and some not.

And it's not clear to me what the starting state is: I'm assuming the first (temporal) thing that happens is a `BEGIN_BLOCK` where both `action.blockHeight` and `savedHeight` are zero, but then the very first thing we do looks like a replay, which seems wrong.

What if, instead of cranking the kernel on each message delivery, we held off cranking it until the `END_BLOCK`? So then all of `BEGIN_BLOCK` and `DELIVER_INBOUND` and `END_BLOCK` would just push actions onto the queue. Then additionally `END_BLOCK` would figure out if we're doing a block for the second time (and compare the contents of the queue against the previously-executed actions), or actually deliver the messages (all at once) and crank the kernel and save the state?

Might that enable a simpler control flow?

Reply:
I can restructure it a bit better (probably a subfunction for each of the actions) and add comments.  I'm just out of cycles this week.

The sequence is:

```
(BEGIN_BLOCK DELIVER_INBOUND* END_BLOCK)+ COMMIT_BLOCK
```

The only reason I avoided the queue-only, crank at `END_BLOCK` design is that it makes it harder to trace the flow of a Go panic to see what transaction caused a crash.

Reply:
The other reason is that transactions aren't billed to the correct gas meter if we run them all in END_BLOCK.